proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,209 @@
+```
+
+Compiled with the `-T cs_6_4` flag this produces the validation error:
+```
+<>:11:3: error: Opcode WriteSamplerFeedbackLevel not valid in shader model cs_6_4.
+note: at 'call void @dx.op.writeSamplerFeedbackLevel(i32 176, %dx.types.Handle %1, %dx.types.Handle %2, %dx.types.Handle %3, float %8, float %9, float undef, float undef, float %11)' in block '#0' of function 'main'.
+Validation failed.
+```
+
+This error displays the LLVM IR text for the error site. In a trivial example
+like the one above finding the error is easy. In a complex shader that could be
+thousands of lines long this is unmanageable.
+
+With this proposal under the _relaxed_ mode, clang will emit the following
+warning:
+
+```
+<>:11:6: warning: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   11 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+With this proposal under the _strict_ or _pedantic_ mode, clang will emit the
+following error:
+
+```
+<>:11:6: error: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   11 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+In all three cases above the DXIL validator will also emit the same diagnostic
+as it does today.
+
+To illustrate the difference between _strict_ and _pedantic_ take the following
+example:
+
+```c++
+Texture2D texture : register(t0);
+SamplerState samplerState : register(s0);
+
+FeedbackTexture2D<SAMPLER_FEEDBACK_MIN_MIP> map : register(u0);
+
+void fn(uint3 threadId) {
+  float2 uv = threadId.xy;
+  uv /= 256;
+  map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+}
+
+[numthreads(4, 4, 1)] 
+void main(uint3 threadId : SV_DispatchThreadId) { }
+```
+
+In this example the call of the unavailable API is in an unused function. In
+both the _relaxed_ and _strict_ diagnostic modes, clang will emit no
+diagnostics. In the pedantic mode, clang emits the following diagnostic:
+
+```
+<>:9:6: error: 'WriteSamplerFeedbackLevel' is available beginning with Shader Model 6.6
+   9 |   map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+     |       ^~~~~~~~~~~~~~~~~~~~~~~~~
+```
+
+In this case the DXIL validator will emit no diagnostic since the call to the
+unavailable API is not present in the final output.
+
+Because bytecode validation occurs late (after optimization), some more complex
+cases can occur. For example, users have reported situations that produce
+validation errors when building un-optimized shaders for debugging proposes that
+do not appear with optimized shaders.
+
+One presentation of such a case is if a call to an unavailable API occurs under
+control flow. If the compiler optimizes away the control flow it may remove the
+API call. In these cases shaders that compile and verify successfully with DXC
+may produce warnings or errors with Clang.

```

Comment:
The example of the difference between strict and pedantic is kind of intermixed with the explanation of how this is different from how things work in dxc. It might be best to put the strict/pedantic example immediately after the definition of pedantic, then have the discussion of the validator and the differences in behaviour between early and late diagnostics.

Reply:
I haven't addressed this comment (yet), although I further expanded examples. I'm thinking about how best to restructure the document.